data/dataexchange/src/main/java/tech/pegasys/teku/data/eraFileFormat/EraFile.java

```diff
@@ -90,18 +105,132 @@ private void getStateIndices() {
     }
     System.out.println("\tState slot: " + stateIndices.getStartSlot());
     System.out.println("\tState index start: " + recordStart);
-    System.out.println("\tOffsets: " + stateIndices.getSlotOffsets());
+  }
+
+  private BeaconState getBeaconState(final ReadEntry entry) throws IOException {
+    final SnappyFramedInputStream is =
+        new SnappyFramedInputStream(new ByteArrayInputStream(entry.getData()));
+
+    return spec.atSlot(stateIndices.getStartSlot())
+        .getSchemaDefinitions()
+        .getBeaconStateSchema()
+        .sszDeserialize(Bytes.of(is.readAllBytes()));
+  }
+
+  private SignedBeaconBlock getBlock(final ReadEntry entry) throws IOException {
+    final SnappyFramedInputStream is =
+        new SnappyFramedInputStream(new ByteArrayInputStream(entry.getData()));
+
+    if (currentSlot == null) {
+      currentSlot = blockIndices.getStartSlot();
+    }
+    return spec.atSlot(currentSlot)
+        .getSchemaDefinitions()
+        .getSignedBeaconBlockSchema()
+        .sszDeserialize(Bytes.of(is.readAllBytes()));
+  }
+
+  private BeaconState verifyStateInArchive() throws IOException {
+    final int offset = stateIndices.getSlotOffsets().get(0);
+    // offset is negative, relative to start of index
+    if (stateIndices.getRecordStart() + offset < 0) {
+      System.out.println("Offset for state goes beyond length of file");
+    }
+    final ReadEntry stateEntry =
+        new ReadEntry(byteBuffer, (int) stateIndices.getRecordStart() + offset);
+    Preconditions.checkArgument(
+        stateEntry.isStateType(), "The first state index doesn't point to state data.");
+    final BeaconState state = getBeaconState(stateEntry);
+    Preconditions.checkArgument(
+        stateIndices.getStartSlot().equals(state.getSlot()),
+        "Reported state slot does not match the stored state.");
+    System.out.println(
+        "State index matches the stored state stored - slot "
+            + state.getSlot()
+            + " ("
+            + state.hashTreeRoot()
+            + ")");
+    return state;
+  }
+
+  private void verifyFile(final SignedBeaconBlock previousArchiveLastBlock) throws IOException {
+    System.out.println("\nVerifying " + filename);
+    final BeaconState verifiedState = verifyStateInArchive();
+    if (!verifiedState.getSlot().isZero()) {
+      verifyBlocksWithReferenceState(verifiedState, previousArchiveLastBlock);
+    }
+  }
+
+  public SignedBeaconBlock getLastBlock() {
+    return lastBlock;
+  }
+
+  private void verifyBlocksWithReferenceState(
+      BeaconState verifiedState, final SignedBeaconBlock previousArchiveLastBlock)
+      throws IOException {
+    currentSlot = blockIndices.getStartSlot();
+    Bytes32 lastRoot = Bytes32.ZERO;
+    int populatedSlots = 0;
+    int emptySlots = 0;
+    for (int i = 0; i < blockIndices.getCount(); i++) {
```

Comment:
maybe moving all to `long` would make CodeQL happy

Reply:
or fix object :) all good